x-pack/libbeat/management/unit.go

```diff
@@ -0,0 +1,400 @@
+
+	expectedCfg := expected.Config
+
+	if expectedCfg == nil {
+		return nil, nil
+	}
+
+	streamStates := make(map[string]unitState, len(expectedCfg.Streams))
+	streamIDs := make([]string, len(expectedCfg.Streams))
+
+	for idx, stream := range expectedCfg.Streams {
+		streamState := unitState{
+			state: status.Unknown,
+			msg:   "",
+		}
+
+		if id := stream.GetId(); id != "" {
+			streamIDs[idx] = id
+			streamStates[id] = streamState
+			continue
+		}
+
+		if cfgName := expectedCfg.GetName(); cfgName != "" {
+			id := fmt.Sprintf("%s.[%d]", cfgName, idx)
+			streamIDs[idx] = id
+			streamStates[id] = streamState
+			continue
+		}
+
+		id := fmt.Sprintf("%s.[%d]", expectedCfg.GetId(), idx)
+		streamIDs[idx] = id
+		streamStates[id] = streamState
+	}
+
+	return streamStates, streamIDs
+}
+
+// newAgentUnit creates a new agentUnit. In case the supplied client.Unit is of type
+// client.UnitTypeInput it initializes the streamStates with a unitState.Unknown
+func newAgentUnit(cu *client.Unit, log *logp.Logger) *agentUnit {
+	var (
+		streamStates map[string]unitState
+		streamIDs    []string
+	)
+
+	if cu.Type() == client.UnitTypeInput {
+		streamStates, streamIDs = getStreamStates(cu.Expected())
+	}
+
+	return &agentUnit{
+		clientUnit:   cu,
+		logger:       log,
+		streamIDs:    streamIDs,
+		streamStates: streamStates,
+	}
+}
+
+// RegisterAction registers action handler for this unit.
+func (u *agentUnit) RegisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.RegisterAction(action)
+}
+
+// UnregisterAction unregisters action handler with the client.
+func (u *agentUnit) UnregisterAction(action client.Action) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return
+	}
+
+	u.clientUnit.UnregisterAction(action)
+}
+
+func (u *agentUnit) Expected() client.Expected {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return client.Expected{}
+	}
+
+	return u.clientUnit.Expected()
+}
+
+func (u *agentUnit) ID() string {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	if u.clientUnit == nil {
+		return ""
+	}
+
+	return u.clientUnit.ID()
+}
+
+// calcState calculates the current state of the unit.
+func (u *agentUnit) calcState() (status.Status, string) {
+	// for type output return the unit state directly as it has no streams
+	if u.clientUnit.Type() == client.UnitTypeOutput {
+		return u.input.state, u.input.msg
+	}
+
+	// if input state is not running return the input state
+	if u.input.state != status.Running {
+		return u.input.state, u.input.msg
+	}
+
+	// input state is marked as running, check the stream states
+	currStatus := status.Running
+	countFailed := 0
+	countDegraded := 0
+	for _, streamState := range u.streamStates {
+		switch streamState.state {
+		case status.Degraded:
+			countDegraded++
+			if currStatus == status.Running {
+				currStatus = status.Degraded
+			}
```

Comment:
You should allow overriding the other states as well so we don't end up with errors hidden behind the starting state.

It seems like you just don't want to convert Failed to Degraded if it was previously failed.

Reply:
>You should allow overriding the other states as well so we don't end up with errors hidden behind the starting state.

To me it doesn't make sense to mark the whole input as Configuring if one stream is doing that but others are Healthy. That said, please do elaborate more on your all-statuses-inclusive overriding logic and I can implement it.

>It seems like you just don't want to convert Failed to Degraded if it was previously failed.

Nope that's not true, every time I check the current status of each stream; notice that currStatus before the loop is set to Running which comes from input level. This if that caught your eye here is to prevent switching the input-level reported status from calcState() as Degraded when you have at least one stream with Failed status. So there is previously Failed.

Reply:
@cmacknz kind reminder for the above

Reply:
> To me it doesn't make sense to mark the whole input as Configuring if one stream is doing that but others are Healthy. 

This is fine but in practice it makes no functional difference, Fleet considers everything except Degraded and Failed as healthy.

The case I care about from the tests is the one below where the input is kept healthy/running when one of the streams is failed (or degraded):

```go
		{
			name:        "one stream failed the other running",
			unit:        unitCfg,
			inputStatus: status.Running,
			streamStates: map[string]status.Status{
				"stream-1": status.Failed,
				"stream-2": status.Running,
			},
			expectedUnitStatus: client.UnitStateFailed,
		},
```

In this case I think the input should be failed with the error of `stream-1` set as its message. The input unit is the primary reporting unit in the agent [architecture](https://github.com/elastic/elastic-agent/blob/main/docs/architecture.md#component-model) and is the entity Fleet knows about today. If some part of the input is failed the whole thing should be failed or at least degraded. That Beats happens to create multiple Beat inputs for each stream is a Beat specific implementation detail. We can include those extra details in the payload, but I want the overall input unit to be an accurate summary what it is doing as a whole.

Reply:
I would also like it if we made the overall input error message include the error of the first degraded/failed as a way of decoupling this change from the UI improvements that should follow. https://github.com/elastic/beats/pull/39209#issuecomment-2123093170